frame/codec.go

```diff
@@ -0,0 +1,136 @@
+
+	// DiscardBody discards the contents of a frame body read from the given source. This is a partial operation; it is
+	// illegal to call this method before calling DecodeHeader.
+	DiscardBody(header *Header, source io.Reader) error
+}
+
+type RawConverter interface {
+
+	// ConvertToRawFrame converts a Frame to a RawFrame, encoding the body and compressing it if necessary. The
+	// returned RawFrame will share the same header with the initial Frame.
+	ConvertToRawFrame(frame *Frame) (*RawFrame, error)
+
+	// ConvertFromRawFrame converts a RawFrame to a Frame, decoding the body and decompressing it if necessary. The
+	// returned Frame will share the same header with the initial RawFrame.
+	ConvertFromRawFrame(frame *RawFrame) (*Frame, error)
+}
+
+// Codec exposes basic encoding and decoding operations for Frame instances. It should be the preferred interface to
+// use in typical client applications such as drivers.
+type Codec interface {
+	Encoder
+	Decoder
+
+	// Returns the BodyCompressor used to compress frame bodies, or nil if none is currently set.
+	GetBodyCompressor() BodyCompressor
+
+	// Sets the BodyCompressor to use to compress messages; passing nil disables compression.
+	// When encoding, in order for a frame body to be compressed, one needs to set a BodyCompressor here and then set
+	// the compression flag in the header of each individual frame to encode, before calling EncodeFrame.
+	// When decoding, the BodyCompressor will be automatically used to decode any frame that has the compression flag
+	// set in its header.
+	// Setting the body compressor may not be a thread-safe operation; it should only be done when initializing
+	// the application, not when the codec is already being used.
+	SetBodyCompressor(compressor BodyCompressor)
+}
+
+// RawCodec exposes advanced encoding and decoding operations for both Frame and RawFrame instances. It should be used
+// only by applications that need to access the frame header without necessarily accessing the frame body, such as
+// proxies or gateways.
+type RawCodec interface {
+	Codec
+	RawEncoder
+	RawDecoder
+	RawConverter
+}
+
+type codec struct {
+	messageCodecs map[primitive.OpCode]message.Codec
+	compressor    BodyCompressor
+}
+
+func NewCodec(messageCodecs ...message.Codec) Codec {
+	return NewRawCodec(messageCodecs...)
+}
+
+func NewRawCodec(messageCodecs ...message.Codec) RawCodec {
+	frameCodec := &codec{
+		messageCodecs: make(map[primitive.OpCode]message.Codec, len(message.DefaultMessageCodecs)+len(messageCodecs)),
+	}
+	for _, messageCodec := range message.DefaultMessageCodecs {
+		frameCodec.messageCodecs[messageCodec.GetOpCode()] = messageCodec
+	}
+	for _, messageCodec := range messageCodecs {
+		frameCodec.messageCodecs[messageCodec.GetOpCode()] = messageCodec
+	}
+	return frameCodec
+}
+
+func (c *codec) GetBodyCompressor() BodyCompressor {
+	return c.compressor
+}
+
+func (c *codec) SetBodyCompressor(compressor BodyCompressor) {
+	c.compressor = compressor
```

Comment:
What about thread safety? This essentially makes the codec not thread safe IIUC. Should we just document that using these two methods makes the codec not thread safe and it is up to the caller to synchronize access to the codec? Or should we recommend that a codec is created every time it is needed instead of reusing it?

Reply:
Yep, I put a note in the docs of `SetBodyCompressor` (at the interface level) to warn about thread safety issues. Indeed I'd prefer to leave the synchronization burden to users because otherwise we would have to synchronize internally every time we want to access the compressor, which would certainly hurt performance.

Reply:
Ah I didn't check the interface, 👍